tchannel-core/src/main/java/com/uber/tchannel/api/SubChannel.java

```diff
@@ -0,0 +1,261 @@
+            final Class<T> responseType,
+            ArgScheme scheme
+    ) throws InterruptedException, TChannelError {
+
+        RawRequest rawRequest = new RawRequest(
+                request.getTTL(),
+                request.getService(),
+                request.getTransportHeaders(),
+                serializer.encodeEndpoint(request.getEndpoint(), scheme),
+                serializer.encodeHeaders(request.getHeaders(), scheme),
+                serializer.encodeBody(request.getBody(), scheme)
+        );
+
+        // Set the "cn" header
+        // TODO: should make "cn" an option
+        rawRequest.setTransportHeader(TransportHeaders.CALLER_NAME_KEY, this.topChannel.getServiceName());
+        rawRequest.setTransportHeader(TransportHeaders.ARG_SCHEME_KEY, scheme.getScheme());
+
+        ListenableFuture<ResponseMessage> future = this.call(host, port, rawRequest);
+        return transform(future, new AsyncFunction<ResponseMessage, Response<T>>() {
+            @Override
+            public ListenableFuture<Response<T>> apply(ResponseMessage responseMessage) {
+                SettableFuture<Response<T>> settableFuture = SettableFuture.create();
+                Response<T> response = null;
+                if (responseMessage instanceof ErrorResponse) {
+                    response = new Response.Builder<T>((ErrorResponse) responseMessage).build();
+                } else {
+                    response = new Response.Builder<>(
+                        serializer.decodeBody((RawResponse) responseMessage, responseType),
+                        serializer.decodeEndpoint((RawResponse) responseMessage),
+                        ((RawResponse) responseMessage).getResponseCode())
+                        .setHeaders(serializer.decodeHeaders((RawResponse) responseMessage))
+                        .build();
+                }
+
+                settableFuture.set(response);
+                return settableFuture;
+            }
+        });
+    }
+
+    public <T, U> ListenableFuture<Response<T>> callThrift(
+            InetAddress host,
+            int port,
+            Request<U> request,
+            final Class<T> responseType
+    ) throws InterruptedException, TChannelError {
+        return callWithEncoding(host, port, request, responseType, ArgScheme.THRIFT);
+    }
+
+    public <T, U> ListenableFuture<Response<T>> callThrift(
+        Request<U> request,
+        final Class<T> responseType
+    ) throws InterruptedException, TChannelError {
+        return callWithEncoding(null, 0, request, responseType, ArgScheme.THRIFT);
+    }
+
+    public <T, U> ListenableFuture<Response<T>> callJSON(
+            InetAddress host,
+            int port,
+            Request<U> request,
+            final Class<T> responseType
+    ) throws InterruptedException, TChannelError {
+        return callWithEncoding(host, port, request, responseType, ArgScheme.JSON);
+    }
+
+    public <T, U> ListenableFuture<Response<T>> callJSON(
```

Comment:
yea, TChannel supports two mode: 1) direct connection to a remote host (already supported in tchannel-java) 2) connections to a list of peers, i.e., peer selection.

The second one is most commonly used in Hyperbahn.